src/java/org/apache/cassandra/utils/JVMStabilityInspector.java

```diff
@@ -140,15 +140,26 @@ else if (t instanceof UnrecoverableIllegalStateException)
             if (t instanceof FSError || t instanceof CorruptSSTableException)
                 isUnstable = true;
 
-        fn.accept(t);
-
         // Check for file handle exhaustion
         if (t instanceof FileNotFoundException || t instanceof FileSystemException || t instanceof SocketException)
             if (t.getMessage() != null && t.getMessage().contains("Too many open files"))
                 isUnstable = true;
 
         if (isUnstable)
+        {
+            if (!StorageService.instance.isDaemonSetupCompleted())
+                FileUtils.handleStartupFSError(t);
             killer.killCurrentJVM(t);
+        }
+
+        try
+        {
+            fn.accept(t);
+        }
+        catch (Exception | Error e)
+        {
+            logger.warn("Unexpected error while handling unexpected error", e);
+        }
```

Comment:
Are the changes in `inspectThrowable()` accounted for in a test anywhere? It doesn't seem to affect any of the tests the patch touches if I move `accept()` back up to its former location...

Reply:
yes, `org.apache.cassandra.utils.JVMStabilityInspectorTest#testKill` triggers this case, It just doesn't validate the warn


```
killerForTests.reset();
JVMStabilityInspector.inspectThrowable(new FSWriteError(new IOException(), "blah"));
assertTrue(killerForTests.wasKilled());
```
that statement triggers this


```
WARN  [main] 2022-02-01 16:52:26,617 JVMStabilityInspector.java:161 - Unexpected error while handling unexpected error
java.lang.IllegalStateException: null
	at org.apache.cassandra.service.DefaultFSErrorHandler.handleFSError(DefaultFSErrorHandler.java:93)
	at org.apache.cassandra.io.util.FileUtils.lambda$handleFSError$1(FileUtils.java:463)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.apache.cassandra.io.util.FileUtils.handleFSError(FileUtils.java:463)
	at org.apache.cassandra.utils.JVMStabilityInspector.inspectDiskError(JVMStabilityInspector.java:100)
	at org.apache.cassandra.utils.JVMStabilityInspector.inspectThrowable(JVMStabilityInspector.java:157)
	at org.apache.cassandra.utils.JVMStabilityInspector.inspectThrowable(JVMStabilityInspector.java:87)
	at org.apache.cassandra.utils.JVMStabilityInspectorTest.testKill(JVMStabilityInspectorTest.java:81)
```